File: browser_ext/test/providers_sync.test.mjs
// sw/bridge.js providers-sync + keyless llm relay tests (issue #34 item 3):
// bridge.js runs in a vm sandbox against a WebSocket stub; hello carries the
// providers-sync capability, sync frames land in chrome.storage.local
// (merged through the Dart agent when present, acked in copy mode), proxy
// mode NEVER lets key bytes touch storage (UT-S1), and sendLlm correlates
// llmRes frames — a dropped link rejects with "desktop link is down" (E26).
import test from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import vm from 'node:vm';
import { webcrypto } from 'node:crypto';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const swDir = join(dirname(fileURLToPath(import.meta.url)), '..', 'sw');
const src = readFileSync(join(swDir, 'bridge.js'), 'utf8');

const tick = () => new Promise((r) => setTimeout(r, 0));

/** Storage stub: local persists into `map`, session is absent (mem fallback). */
function storageStub(map) {
  return {
    local: {
      async get(keys) {
        const out = {};
        for (const k of [].concat(keys ?? [])) if (k in map) out[k] = map[k];
        return out;
      },
      async set(o) { Object.assign(map, JSON.parse(JSON.stringify(o))); },
      async remove(keys) { for (const k of [].concat(keys)) delete map[k]; },
    },
  };
}

/** Sandbox with a WebSocket stub; every sent frame is recorded in `sent`. */
function boot({ faAgent } = {}) {
  const map = {}; // chrome.storage.local backing
  const sent = [];
  const sockets = [];
  class WsStub {
    constructor(url) {
      this.url = url;
      this.readyState = 0;
      sockets.push(this);
    }
    send(data) { sent.push(JSON.parse(data)); }
    close() { this.readyState = 3; }
  }
  const sandbox = {
    console,
    crypto: webcrypto,
    WebSocket: WsStub,
    chrome: { storage: storageStub(map), alarms: { create() {} } },
    setTimeout,
    clearTimeout,
    // The ping keepalive would hold the node process open for its full
    // 20s interval after the tests finish — stub it out; ping timing is
    // not under test here.
    setInterval: () => 0,
    clearInterval: () => {},
    Date,
    JSON,
    Promise,
    Error,
    Map,
    Set,
    Array,
    Object,
  };
  if (faAgent) sandbox.faAgent = faAgent;
  vm.createContext(sandbox);
  vm.runInContext(src, sandbox, { filename: 'bridge.js' });
  return { sandbox, map, sent, sockets };
}

/** Feed one wire frame the way the browser would (MessageEvent.data). */
const feed = (ws, frame) => ws.onmessage({ data: JSON.stringify(frame) });

/** Pair + finish the handshake; returns the open socket. */
async function connected({ sandbox, sockets }, url = 'ws://127.0.0.1:8777/ws') {
  await sandbox.faSw.bridge.connect(url, 'tok');
  const ws = sockets.at(-1);
  ws.readyState = 1;
  await ws.onopen();
  feed(ws, { v: 1, id: 'w1', op: 'welcome', mailbox: 'mb' });
  return ws;
}

test('hello advertises the providers-sync capability', async () => {
  const ctx = boot();
  await ctx.sandbox.faSw.bridge.connect('ws://x', 'tok');
  const ws = ctx.sockets.at(-1);
  ws.readyState = 1;
  await ws.onopen();
  const hello = ctx.sent.find((f) => f.op === 'hello');
  assert.ok(hello, 'hello sent');
  assert.ok(hello.caps.includes('providers-sync'), 'caps include providers-sync');
});

test('copy-mode sync: merged doc stored, keys ride once, acked echoes the sync frame id', async (t) => {
  const ctx = boot({
    faAgent: {
      // Stand-in for the Dart providersMerge: keeps the local edit alive.
      async providersMerge(prev, next) {
        t.diagnostic?.('merge called');
        ctx.mergedWith = { prev, next };
        return {
          ...next,
          providers: [
            ...next.providers,
            { name: 'mine', baseUrl: 'http://x', modelId: 'm', provenance: 'local' },
          ],
        };
      },
    },
  });
  const ws = await connected(ctx);
  const syncId = 'sync-42';
  feed(ws, {
    v: 1, id: syncId, op: 'providersSync',
    sync: {
      version: 1, mode: 'copy', host: 'desk',
      providers: [{ name: 'p1', apiType: 'openai', baseUrl: 'https://a/v1', modelId: 'gpt', provenance: 'synced-from-cli@desk' }],
      keys: { p1: 'sk-COPY-SECRET' },
    },
  });
  await tick();
  const doc = ctx.map.faProviders;
  assert.equal(doc.mode, 'copy');
  assert.equal(doc.keys.p1, 'sk-COPY-SECRET', 'copy keys land in storage');
  assert.ok(doc.providers.some((p) => p.name === 'mine'), 'local entry survived merge (UT-S2)');
  assert.equal(ctx.mergedWith.prev, undefined, 'no prior doc passed');
  const ack = ctx.sent.find((f) => f.op === 'acked');
  assert.ok(ack, 'copy mode acked');
  assert.equal(ack.id, syncId, 'ack echoes the sync frame id');
});

test('re-pair without the agent (scaffold SW): raw sync lands as-is, still acked in copy mode', async () => {
  const ctx = boot(); // no faAgent
  const ws = await connected(ctx);
  feed(ws, {
    v: 1, id: 's2', op: 'providersSync',
    sync: { version: 1, mode: 'copy', host: 'desk', providers: [{ name: 'p1', baseUrl: 'https://a', modelId: 'm', provenance: 'synced-from-cli@desk' }], keys: { p1: 'k' } },
  });
  await tick();
  assert.equal(ctx.map.faProviders.providers.length, 1);
  assert.ok(ctx.sent.some((f) => f.op === 'acked' && f.id === 's2'));
});

test('UT-S1: proxy-mode keys NEVER touch storage (byte-scan over everything stored)', async () => {
  const ctx = boot();
  const ws = await connected(ctx);
  feed(ws, {
    v: 1, id: 's3', op: 'providersSync',
    sync: {
      version: 1, mode: 'proxy', host: 'desk',
      providers: [{ name: 'p1', apiType: 'openai', baseUrl: 'https://a/v1', modelId: 'gpt', provenance: 'synced-from-cli@desk' }],
      keys: { p1: 'sk-BYTESCAN-SECRET-xyz' },
    },
  });
  await tick();
  const everything = JSON.stringify(ctx.map);
  assert.ok(!everything.includes('sk-BYTESCAN-SECRET-xyz'), 'no key byte anywhere in storage');
  assert.equal(ctx.map.faProviders.mode, 'proxy');
  assert.equal(ctx.map.faProviders.keys, undefined, 'keys field dropped');
  assert.ok(ctx.map.faProviders.providers[0].provenance.startsWith('synced-from-cli@'));
  assert.ok(!ctx.sent.some((f) => f.op === 'acked'), 'proxy mode does not ack');
});

test('sendLlm streams deltas over llmReq/llmRes and resolves on done', async () => {
  const ctx = boot();
  const ws = await connected(ctx);
  const deltas = [];
  const done = ctx.sandbox.faSw.bridge.sendLlm(
    { baseUrl: 'https://a/v1', model: 'gpt', messages: [{ role: 'user', content: 'hi' }] },
    (d) => deltas.push(d),
  );
  await tick();
  const req = ctx.sent.find((f) => f.op === 'llmReq');
  assert.ok(req, 'llmReq sent');
  assert.equal(req.req.model, 'gpt');
  const id = req.id;
  feed(ws, { v: 1, id, op: 'llmRes', delta: 'he' });
  feed(ws, { v: 1, id, op: 'llmRes', delta: 'llo' });
  feed(ws, { v: 1, id, op: 'llmRes', done: true });
  await done;
  assert.deepEqual(deltas, ['he', 'llo']);
});

test('sendLlm rejects on an error frame and ignores frames for unknown ids', async () => {
  const ctx = boot();
  const ws = await connected(ctx);
  const p = ctx.sandbox.faSw.bridge.sendLlm({ model: 'm', messages: [] }, () => {});
  await tick();
  const id = ctx.sent.find((f) => f.op === 'llmReq').id;
  feed(ws, { v: 1, id: 'other', op: 'llmRes', done: true }); // unknown id — no-op
  feed(ws, { v: 1, id, op: 'llmRes', error: 'upstream 401' });
  await assert.rejects(p, /upstream 401/);
});

test('E26: sendLlm while disconnected rejects "desktop link is down"; a mid-stream drop rejects cleanly', async () => {
  const ctx = boot();
  await assert.rejects(
    ctx.sandbox.faSw.bridge.sendLlm({ model: 'm', messages: [] }, () => {}),
    /desktop link is down/,
  );
  const ws = await connected(ctx);
  const p = ctx.sandbox.faSw.bridge.sendLlm({ model: 'm', messages: [] }, () => {});
  await tick();
  ws.onclose(); // link drops mid-stream
  await assert.rejects(p, /desktop link is down/);
});

File: browser_ext/dart/test/providers_sync_test.dart
// Providers registry merge/routing tests (issue #34 item 3): pure Dart.
//
// UT-S2 — re-pair overwrites synced entries but NEVER silently clobbers
// local edits. E26 — a keyless synced entry routes through the relay and
// a dead relay surfaces "desktop link is down" as a clean stream error.
library;

import 'dart:io';

import '../src/providers.dart';
import 'package:flutter_agent_harness/src/cancel_token.dart';
import 'package:flutter_agent_harness/src/context.dart';
import 'package:flutter_agent_harness/src/event_stream.dart';
import 'package:flutter_agent_harness/src/model.dart';
import 'package:flutter_agent_harness/src/types.dart';
import 'package:test/test.dart';

const synced = 'synced-from-cli@desk';

Map<String, dynamic> doc(
  List<Map<String, dynamic>> providers, {
  String mode = 'proxy',
}) => {'version': 1, 'mode': mode, 'host': 'desk', 'providers': providers};

Map<String, dynamic> entryJson(
  String name, {
  String provenance = synced,
  String apiKey = '',
  String modelId = 'gpt-x',
  String baseUrl = 'https://api/v1',
}) => {
  'name': name,
  'apiType': 'openai',
  'baseUrl': baseUrl,
  'modelId': modelId,
  'provenance': provenance,
  if (apiKey.isNotEmpty) 'apiKey': apiKey,
};

ProviderEntry entry(
  String name, {
  String provenance = synced,
  String apiKey = '',
  String modelId = 'gpt-x',
  String baseUrl = 'https://api/v1',
}) => (
  name: name,
  apiType: 'openai',
  baseUrl: baseUrl,
  modelId: modelId,
  provenance: provenance,
  apiKey: apiKey,
);

void main() {
  test('mergeSyncedProviders replaces synced entries and keeps local ones', () {
    final existing = doc([
      entryJson('cli-one'),
      entryJson('mine', provenance: localProvenance, apiKey: 'sk-local'),
    ]);
    final merged = mergeSyncedProviders(existing, {
      'version': 1,
      'mode': 'copy',
      'host': 'desk',
      'providers': [
        entryJson('cli-one', modelId: 'gpt-new'),
        entryJson('cli-two'),
      ],
      'keys': {'cli-one': 'sk-staged'},
    });

    final byName = {
      for (final p in (merged['providers'] as List).cast<Map>())
        p['name'] as String: p,
    };
    // UT-S2: the local edit survives a re-pair untouched.
    expect(byName['mine']!['apiKey'], 'sk-local');
    expect(byName['mine']!['provenance'], localProvenance);
    // Synced rows are replaced wholesale; copy keys attach.
    expect(byName['cli-one']!['modelId'], 'gpt-new');
    expect(byName['cli-one']!['apiKey'], 'sk-staged');
    // Keyless synced entries omit the field entirely — never a stale key.
    expect(byName['cli-two']!['apiKey'] ?? '', '');
    expect(merged['mode'], 'copy');
  });

  test('mergeSyncedProviders on a null doc lands the sync as-is (proxy)', () {
    final merged = mergeSyncedProviders(null, {
      'version': 1,
      'mode': 'proxy',
      'host': 'desk',
      'providers': [entryJson('p1')],
    });
    expect(merged['mode'], 'proxy');
    expect((merged['providers'] as List).length, 1);
  });

  test(
    'mergeLocalProviders stamps local provenance and replaces same-name',
    () {
      final existing = doc([entryJson('p1')]);
      final merged = mergeLocalProviders(existing, [
        (
          name: 'p1',
          apiType: 'openai',
          baseUrl: 'https://imported/v1',
          modelId: 'imported-model',
          provenance: localProvenance,
          apiKey: 'sk-imported',
        ),
        (
          name: 'added',
          apiType: 'openai',
          baseUrl: 'https://b/v1',
          modelId: 'm2',
          provenance: localProvenance,
          apiKey: 'k2',
        ),
      ]);
      final byName = {
        for (final p in (merged['providers'] as List).cast<Map>())
          p['name'] as String: p,
      };
      // An explicit local import wins over a synced row.
      expect(byName['p1']!['baseUrl'], 'https://imported/v1');
      expect(byName['p1']!['provenance'], localProvenance);
      expect(byName['added']!['provenance'], localProvenance);
      expect((merged['providers'] as List).length, 2);
    },
  );

  test('removeProvider drops one entry by name', () {
    final removed = removeProvider(
      doc([entryJson('a'), entryJson('b', provenance: localProvenance)]),
      'a',
    );
    final names = (removed['providers'] as List).map((p) => p['name']);
    expect(names, ['b']);
  });

  test('pickActiveEntry: registry is primary, legacy map is the fallback', () {
    expect(
      pickActiveEntry(
        legacy: {'baseUrl': 'https://legacy', 'model': 'legacy-model'},
        doc: doc([entryJson('p1')]),
      )!.name,
      'p1',
    );
    final legacy = pickActiveEntry(
      legacy: {'baseUrl': 'https://legacy', 'apiKey': 'k', 'model': 'm'},
    );
    expect(legacy!.baseUrl, 'https://legacy');
    expect(pickActiveEntry(legacy: null, doc: null), isNull);
  });

  test(
    'routeFor: fake beats everything; keyless synced = relay; else direct',
    () {
      expect(routeFor(entry('f', modelId: 'fake:echo')), ProviderRoute.fake);
      expect(routeFor(entry('r', apiKey: '')), ProviderRoute.relay);
      expect(routeFor(entry('d', apiKey: 'sk-1')), ProviderRoute.direct);
      // Keyless LOCAL endpoints (Ollama/llama.cpp) fetch directly.
      expect(
        routeFor(entry('l', apiKey: '', provenance: localProvenance)),
        ProviderRoute.direct,
      );
    },
  );

  test(
    'E26: a dead relay surfaces "desktop link is down" as a clean error',
    () async {
      final events = relayTextStream(
        modelForConfig((baseUrl: 'https://api/v1', apiKey: '', model: 'gpt-x')),
        // The JS relay rejects with exactly this message (bridge.js E26);
        // relayTextStream must surface it as a clean ErrorEvent, not throw.
        Stream<String>.error(const SocketException('desktop link is down')),
      );
      final terminal = await events.last;
      expect(terminal, isA<ErrorEvent>());
      final error = await events.result;
      expect(error.stopReason, StopReason.error);
      expect(error.errorMessage, contains('desktop link is down'));
    },
  );

  test('relayTextStream streams deltas then DoneEvent', () async {
    final events = relayTextStream(
      modelForConfig((baseUrl: 'b', apiKey: '', model: 'm')),
      Stream<String>.fromIterable(['he', 'llo']),
    );
    final terminals = <AssistantMessageEvent>[];
    await for (final e in events) {
      terminals.add(e);
    }
    expect(terminals.first, isA<TextDeltaEvent>());
    expect((terminals.first as TextDeltaEvent).delta, 'he');
    expect(terminals.last, isA<DoneEvent>());
    expect(
      ((terminals.last as DoneEvent).message.content.single as TextContent)
          .text,
      'hello',
    );
  });

  test('resolveStreamFn: relay-routed active config uses the relay', () {
    final relayStreams = <Stream<String>>[];
    final fakeRelay = _FakeRelay(relayStreams);
    activeRelay = (relay: fakeRelay, providerName: 'p1');
    addSetterTearDown(() => activeRelay = null);

    final keyless = (baseUrl: 'https://api/v1', apiKey: '', model: 'gpt-x');
    var fn = resolveStreamFn(keyless);
    // The resolved function must call the relay, not fetch.
    fn(
      modelForConfig(keyless),
      Context(systemPrompt: null, messages: [UserMessage.text('hi')]),
    );
    expect(fakeRelay.calls, 1);

    // A keyed config ignores the installed relay.
    resolveStreamFn((baseUrl: 'https://api/v1', apiKey: 'sk', model: 'gpt-x'));
    expect(fakeRelay.calls, 1);

    // fake: stays scripted.
    expect(
      identical(
        resolveStreamFn((baseUrl: '', apiKey: '', model: 'fake:echo')),
        fakeStream,
      ),
      isTrue,
    );
  });

  test('cookie-auth hosts bypass the relay and stream direct', () {
    final relayStreams = <Stream<String>>[];
    final fakeRelay = _FakeRelay(relayStreams);
    activeRelay = (relay: fakeRelay, providerName: 'p1');
    addSetterTearDown(() => activeRelay = null);

    // A keyless CodeMie entry must NOT ride the relay — the relay would
    // strip the browser cookies and CodeMie would 401. The direct SW
    // fetch (FetchClient credentials:include) carries the jar.
    final codemie = (
      baseUrl: 'https://codemie.lab.epam.com/code-assistant-api/v1',
      apiKey: '',
      model: 'gpt-x',
    );
    expect(isCookieAuthUrl(codemie.baseUrl), isTrue);
    final fn = resolveStreamFn(codemie);
    fn(
      modelForConfig(codemie),
      Context(systemPrompt: null, messages: [UserMessage.text('hi')]),
    );
    expect(fakeRelay.calls, 0, reason: 'cookie-auth host streams direct');

    // A non-CodeMie host keeps the relay routing.
    final relayed = resolveStreamFn((
      baseUrl: 'https://api/v1',
      apiKey: '',
      model: 'gpt-x',
    ));
    relayed(
      modelForConfig((baseUrl: 'https://api/v1', apiKey: '', model: 'gpt-x')),
      Context(systemPrompt: null, messages: [UserMessage.text('hi')]),
    );
    expect(fakeRelay.calls, 1);
  });

  test('isCookieAuthUrl: classification edges', () {
    expect(isCookieAuthUrl('https://codemie.lab.epam.com/'), isTrue);
    expect(isCookieAuthUrl('https://api.openai.com/v1'), isFalse);
    expect(isCookieAuthUrl('not a url'), isFalse);
    expect(isCookieAuthUrl(''), isFalse);
  });
}

void addSetterTearDown(void Function() fn) => addTearDown(fn);

class _FakeRelay implements BridgeLlmRelay {
  _FakeRelay(this.streams);

  final List<Stream<String>> streams;
  int calls = 0;

  @override
  bool get connected => true;

  @override
  AssistantMessageEventStream stream(
    Model model,
    Context context, {
    CancelToken? cancelToken,
    String? providerName,
  }) {
    calls++;
    return relayTextStream(model, Stream<String>.empty());
  }
}
